Add access key shortcuts to Header, Input, and ModelSelector components

chi1180 opened the pull request:
- Header: Added access keys for new conversation (n), history (h), and delete (d).
- Input: Added access key (m) for the message input field.
- ModelSelector: Added access key (/) for the model selector dropdown.

When user press [alt] + [access key], th element will be focused

chi1180 committed:
Add access key shortcuts to Header, Input, and ModelSelector components

- Header: Added access keys for new conversation (n), history (h), and
  delete (d).
- Input: Added access key (m) for the message input field.
- ModelSelector: Added access key (/) for the model selector dropdown.

When user press [alt] + [access key], th element will be focused

Pierrad commented:
Hi @chi1180, thanks a lot for your contribution!

It seems interesting, before merging it, is it possible for you to run a lint on the modified files?
Indentation and file structure seems to be changed a lot and does not match other files of the project.

Thanks in advance!

chi1180 committed:
Add keyboard shortcuts for chat actions with react-hotkeys-hook

chi1180 commented:
- Implemented with `react-hotkeys-hook` instead of `accessKey` for the best practice.
- Added `format` command in the `package.json`

Pierrad commented:
Hi @chi1180, I've tested your feature and it works great for the most part. I just had issues with the Model Selector and the input, I pressed the configured keys but it did not focus on the selector neither the input box. 
I changed the hot keys to something else and it was working (maybe overlapping hot keys config on my side). 

This issue made me think that it would be great to be able to choose in the plugin settings the hot keys that we want to use like I initially did with the copilot suggestion system. With this, we will avoid future feedback of users saying that hot keys do not work. 

Do you think you can implement this? :)

chi1180 commented:
Sure !
Thank you for your review !

chi1180 commented:
Sure! Thank you for your review!

Here's the plan to make the chat hotkeys configurable through the plugin settings, following the same pattern already used for the inline copilot keybindings:

---

### 1. Extend the settings types (`src/settings/CopilotPluginSettingTab.tsx`)

Add a `ChatHotkeys` type and a `chatHotkeys` field to `CopilotPluginSettings`, with sensible defaults:

```typescript
export type ChatHotkeys = {
  newConversation: string;       // default: "alt+n"
  conversationHistory: string;   // default: "alt+h"
  deleteConversation: string;    // default: "alt+d"
  focusInput: string;            // default: "alt+m"
  focusModelSelector: string;    // default: "alt+shift+m"
};
```

And in `DEFAULT_SETTINGS`:
```typescript
chatHotkeys: {
  newConversation: "alt+n",
  conversationHistory: "alt+h",
  deleteConversation: "alt+d",
  focusInput: "alt+m",
  focusModelSelector: "alt+shift+m",
},
```

---

### 2. Add a "Chat Keybindings" section to the settings UI

Reuse the existing `KeybindingInput` component (already used for inline copilot keybindings) and render a new section under "Copilot Chat Settings":

```tsx
const chatBindings = [
  {
    title: "New conversation",
    description: "Hotkey to start a new chat conversation.",
    value: this.plugin.settings.chatHotkeys.newConversation,
    onChange: (value: string) => { this.plugin.settings.chatHotkeys.newConversation = value; },
    defaultValue: DEFAULT_SETTINGS.chatHotkeys.newConversation,
  },
  // ... same for the other 4 actions
];
```

---

### 3. Read hotkeys from settings in each component

In `Header.tsx`, `Input.tsx`, and `ModelSelector.tsx`, access the configured hotkeys via the `usePlugin()` hook (which gives access to `plugin.settings`) and pass them to `useHotkeys`:

```tsx
const plugin = usePlugin();
const chatHotkeys = plugin?.settings?.chatHotkeys ?? DEFAULT_SETTINGS.chatHotkeys;

useHotkeys(chatHotkeys.newConversation, handleNewConversation, {
  description: "Start new conversation",
});
useHotkeys(chatHotkeys.conversationHistory, toggleConversationSelector, {
  description: "View conversation history",
});
useHotkeys(chatHotkeys.deleteConversation, handleClearChat, {
  description: "Delete this conversation",
});
```

The `aria-label` and `title` hints on buttons and the textarea placeholder can either be made dynamic (showing the configured key) or simplified to remove the hardcoded key combos so they don't become stale when the user changes the binding.

---

### 4. Note on `KeybindingInput` format vs. `react-hotkeys-hook` format

The existing `KeybindingInput` component stores hotkeys in a CodeMirror-style format (e.g. `"Alt-n"`, `"Cmd-Shift-M"`), while `react-hotkeys-hook` expects lowercase keys with `+` as separator (e.g. `"alt+n"`, `"alt+shift+m"`). You'll need a small conversion utility, or you can store the value in `react-hotkeys-hook` format and adjust the `KeybindingInput` modifier map to use `+` separators and lowercase modifiers.

---

Would you like me to implement these changes directly in this branch?

chi1180 committed:
Make chat hotkeys configurable via plugin settings

- Add ChatHotkeys type and chatHotkeys field to CopilotPluginSettings
  with defaults: alt+n (new), alt+h (history), alt+d (delete),
  alt+m (focus input), alt+shift+m (focus model selector)
- Add 'Chat Keybindings' section in the settings UI under
  'Copilot Chat Settings', reusing the KeybindingInput component
- Update KeybindingInput with an optional format prop ('codemirror'
  or 'hotkeys-hook') so chat bindings are stored in react-hotkeys-hook
  format (e.g. alt+n) while inline copilot bindings remain unchanged
- Read chatHotkeys from plugin settings in Header, Input, and
  ModelSelector; fall back to DEFAULT_SETTINGS if unavailable
- aria-labels, title attrs and the textarea placeholder now reflect
  the configured key combo dynamically

chi1180 commented:
I've implemented configurable chat hotkeys in commit ecd1e6d. Here's a summary of what changed:

### New: `ChatHotkeys` type + `chatHotkeys` in settings

`CopilotPluginSettings` now has a `chatHotkeys` field with these defaults:

| Action | Default key |
|---|---|
| New conversation | `alt+n` |
| Conversation history | `alt+h` |
| Delete conversation | `alt+d` |
| Focus message input | `alt+m` |
| Focus model selector | `alt+shift+m` |

Settings are persisted the same way as all other plugin settings.

### New: "Chat Keybindings" section in the settings UI

A new **Chat Keybindings** section appears under **Copilot Chat Settings**, using the existing `KeybindingInput` component. Users click a binding field, press their desired key combination, then click **Save chat keybindings**.

### Updated `KeybindingInput` — new `format` prop

The component now accepts an optional `format` prop (`"codemirror"` | `"hotkeys-hook"`). 
- The existing inline copilot bindings continue to use `"codemirror"` format (e.g. `Cmd-Shift-/`) — **no breaking change**.  
- The new chat bindings use `"hotkeys-hook"` format (e.g. `alt+shift+m`), which is what `react-hotkeys-hook` expects.

Also improved: the input field is now cleared when focused so users can type a fresh binding without having to clear first.

### Dynamic labels

`aria-label` on header buttons, the `title` on the model selector `<select>`, and the textarea `placeholder` all reflect the currently configured key combo at render time, so they stay in sync when the user changes the bindings.

chi1180 committed:
Fix hotkeys not firing when textarea is focused

react-hotkeys-hook suppresses hotkeys inside form elements (textarea,
input, select) by default. Add enableOnFormTags: true to all five
chat hotkey registrations so they fire regardless of which element
currently has focus.

chi1180 committed:
Fix KeybindingInput blocking all keyboard input in settings UI

Two bugs caused other settings inputs to be unfocusable:

1. event.preventDefault() was called unconditionally for every keydown
   while keyListenerActive was true. This blocked Tab (and other
   navigation keys) from moving focus between setting fields, even
   when the event originated from a completely different input element,
   because the listener was on window.
   Fix: skip preventDefault (and skip recording) for Tab and Escape.

2. handleInputFocus called onChange('') immediately on focus, saving
   an empty string to settings even if the user only tabbed through
   without intending to change the binding.
   Fix: introduce a pendingFirstKey flag — the existing value is kept
   and shown until the user actually presses a key, at which point the
   accumulator starts fresh from an empty base.
